File: benchmarks/matmul/nim/matmul.nim
import std/strutils
import std/os

type F64Arr = ptr UncheckedArray[float64]

proc splitmix64(x: uint64): uint64 =
  var z = x + 0x9E3779B97F4A7C15'u64
  z = (z xor (z shr 30)) * 0xBF58476D1CE4E5B9'u64
  z = (z xor (z shr 27)) * 0x94D049BB133111EB'u64
  z = z xor (z shr 31)
  return z

proc fill(mat: F64Arr, n: int32, seed: uint64) =
  for i in 0 ..< n:
    for j in 0 ..< n:
      let idx = uint64(i * n + j + 1) + seed
      mat[i * n + j] = float64(splitmix64(idx) mod 1000003'u64)

proc matmul(a: F64Arr, b: F64Arr, c: F64Arr, n: int32) =
  for i in 0 ..< n:
    for j in 0 ..< n:
      c[i * n + j] = 0.0
    for k in 0 ..< n:
      let aik = a[i * n + k]
      for j in 0 ..< n:
        c[i * n + j] += aik * b[k * n + j]

proc checksum(mat: F64Arr, n: int32): float64 =
  var sum: float64 = 0.0
  let total = n * n
  for i in 0 ..< total:
    sum += mat[i]
  return sum

when isMainModule:
  if paramCount() < 1:
    stderr.writeLine("Usage: matmul <n>")
    quit(1)
  let n = parseInt(paramStr(1)).int32

  if n <= 0:
    echo "0.000000"
    quit(0)

  let total = n * n
  let a = cast[F64Arr](alloc(total * sizeof(float64)))
  let b = cast[F64Arr](alloc(total * sizeof(float64)))
  let c = cast[F64Arr](alloc(total * sizeof(float64)))

  fill(a, n, 0)
  fill(b, n, 0x9E3779B9'u64)

  matmul(a, b, c, n)

  let result = checksum(c, n)
  dealloc(a)
  dealloc(b)
  dealloc(c)

  echo formatFloat(result, ffDecimal, 6)

File: benchmarks/matmul/zig/matmul.zig
const std = @import("std");

fn splitmix64(x: u64) u64 {
    var z: u64 = x +% 0x9E3779B97F4A7C15;
    z = (z ^ (z >> 30)) *% 0xBF58476D1CE4E5B9;
    z = (z ^ (z >> 27)) *% 0x94D049BB133111EB;
    z = z ^ (z >> 31);
    return z;
}

fn fill(mat: []f64, n: i32, seed: u64) void {
    var i: i32 = 0;
    while (i < n) : (i += 1) {
        var j: i32 = 0;
        while (j < n) : (j += 1) {
            const idx: u64 = @intCast(i * n + j + 1);
            const h = splitmix64(idx +% seed) % 1000003;
            mat[@intCast(i * n + j)] = @floatFromInt(h);
        }
    }
}

fn matmul(a: []const f64, b: []const f64, c: []f64, n: i32) void {
    var i: i32 = 0;
    while (i < n) : (i += 1) {
        var j: i32 = 0;
        while (j < n) : (j += 1) {
            c[@intCast(i * n + j)] = 0.0;
        }
        var k: i32 = 0;
        while (k < n) : (k += 1) {
            const aik = a[@intCast(i * n + k)];
            var j2: i32 = 0;
            while (j2 < n) : (j2 += 1) {
                c[@intCast(i * n + j2)] += aik * b[@intCast(k * n + j2)];
            }
        }
    }
}

fn checksum(mat: []const f64, n: i32) f64 {
    var sum: f64 = 0.0;
    const total = n * n;
    var i: i32 = 0;
    while (i < total) : (i += 1) {
        sum += mat[@intCast(i)];
    }
    return sum;
}

pub fn main(init: std.process.Init) !void {
    var it = init.minimal.args.iterate();
    _ = it.skip(); // program name

    const n_str = it.next() orelse {
        std.debug.print("Usage: matmul <n>\n", .{});
        std.process.exit(1);
    };

    const n = try std.fmt.parseInt(i32, n_str, 10);

    var buf: [64]u8 = undefined;
    var stdout = std.Io.File.stdout().writer(init.io, &buf);

    if (n <= 0) {
        try stdout.interface.print("0.000000\n", .{});
        try stdout.interface.flush();
        return;
    }

    const total: usize = @intCast(n * n);
    const allocator = std.heap.page_allocator;
    const a = try allocator.alloc(f64, total);
    const b = try allocator.alloc(f64, total);
    const c = try allocator.alloc(f64, total);
    defer allocator.free(a);
    defer allocator.free(b);
    defer allocator.free(c);

    fill(a, n, 0);
    fill(b, n, 0x9E3779B9);

    matmul(a, b, c, n);

    const result = checksum(c, n);

    try stdout.interface.print("{d:.6}\n", .{result});
    try stdout.interface.flush();
}
